### DISCUSSION TITLE
Security keys 2FA

### DISCUSSION THREAD

**Discussion by MasterInQuestion** [MasterInQuestion/talk#38] [Informatics]
[[
<b id="discussioncomment-11035416">[ MasterInQuestion @ CE 2024-10-24 01:53:34 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11035416](https://github.com/orgs/community/discussions/22500#discussioncomment-11035416)</b>
    “Very strong password are not sufficient anymore for great account security.
    Today, too many accounts are being hacked on all websites, even if they have strong passwords because there are much more powerful hacking techniques than just brut forcing their passwords.”
    Password itself may function as if security key:
    "/dev/random" derived, unique for each site.
    When properly managed: much identical. (and likely more reliable)

    See also: https://github.com/orgs/community/discussions/63813#discussioncomment-10626201
\
\
    It appears many people have fundamental flaws on the understanding of security:
    Thoughtlessly phasing things out in vain hope, for feigned sense of "security".
    But failed to realize:
    “Security doesn't itself come by following random security advices or blindly deploying "security-enhanced" setup:
    Well realizing the underlying implementation is the essence to guarantee effectiveness and security.”
    .
    https://github.com/orgs/community/discussions/22814#discussioncomment-7571079 <b>]</b>

\----

<b id="discussioncomment-11036997">[ Explorer09 @ CE 2024-10-24 06:37:50 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11036997](https://github.com/orgs/community/discussions/22500#discussioncomment-11036997)</b>
    Why are we believing in a bulls**t that "Password itself may function as if security key"?

    Passwords, no matter how random or secure you generated them, are the same tokens every time you log in.
    Passwords are vulnerable to [replay attacks](https://en.wikipedia.org/wiki/Replay_attack). That was the reason for 2FA in the first place.
    Applications that have to stick with password authentication are shifted to use [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) as a password alternative.
    Personal access tokens can expire after a set days, and have limited privileges per token: and thus reduce the risks of what passwords can do. <b>]</b>

\----

<b id="discussioncomment-11047758">[ MasterInQuestion @ CE 2024-10-25 03:51:23 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11047758](https://github.com/orgs/community/discussions/22500#discussioncomment-11047758)</b>
    Before bold assertion: do proper research.
    Why would you fallaciously believe that random keys each time would miraculously increase security?
    For some inconsiderate genius coined that the ideal encryption model would be alike, given the "absolute reliable" key-exchange channel..?

    For really security-critical applications: 1 compromise already means done.
    Reusable or not is already passé.

    Your thoughts are exactly like those who coined security models that don't work at all:
    Secure or not: much in question.
    Functionality-wise: everything broken.

    Your quoth of Replay Attack is also not much appropriate:
    But whatsoever doesn't matter much. <b>]</b>

\----

<b id="discussioncomment-11049564">[ Explorer09 @ CE 2024-10-25 08:02:58 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11049564](https://github.com/orgs/community/discussions/22500#discussioncomment-11049564)</b>
    "Why would you fallaciously believe that random keys each time would miraculously increase security?"
    I don't. I am just refuting your claim about passwords being secure enough.
    If what you claim is true, why the heck do websites force 2FA on us? <b>]</b>

\----

<b id="discussioncomment-11051733">[ MasterInQuestion @ CE 2024-10-25 11:46:16 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11051733](https://github.com/orgs/community/discussions/22500#discussioncomment-11051733)</b>
    Thoughtless assertion that users are all thoughtlessly using literally "password" as password.
    Some did alike, indeed.
    But such apparently don't care about security.
    Mostly nor capable of maintaining security.
    Critical business demands operator competence: an unavoidable fact.

    Good faith assumed, however:
    Some just excuse this for data collection. <b>]</b>

\----

<b id="discussioncomment-11051892">[ Explorer09 @ CE 2024-10-25 11:50:58 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11051892](https://github.com/orgs/community/discussions/22500#discussioncomment-11051892)</b>
    There are a lot of password strength checkers you can find on the web.
    Please don't assume we know nothing about password strength.
    [https://www.security.org/how-secure-is-my-password/](https://www.security.org/how-secure-is-my-password/)
    [https://bitwarden.com/password-strength](https://bitwarden.com/password-strength)
    [https://www.passwordmonster.com/](https://www.passwordmonster.com/)

    For your argument about data collection, yes.
    This is my own reason not to use 2FA approach other than the security key:
    SMS based 2FA leaks your phone number.
    App based 2FA (including TOTP) requires you to trust the app - something can be programmed to spy on you.
    Hardware security keys stay offline until you need them for authentication: so it is the best 2FA approach for people who have privacy concerns. <b>]</b>

\----

<b id="discussioncomment-11052259">[ MasterInQuestion @ CE 2024-10-25 12:32:50 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11052259](https://github.com/orgs/community/discussions/22500#discussioncomment-11052259)</b>
    I don't. Find who force 2FA to blame.
    Hardware-based are not necessarily safer than software.
    And actually hardware and software cannot really separate...

    The key benefit of security keys alike: is the key generation core being not easily copiable from outside.
    In some cases this may of help.
    But do note: operating security keys in untrusted environment still presents the risk of session hijacking. <b>]</b>

\----

<b id="discussioncomment-11052347">[ lriley2020 @ CE 2024-10-25 12:47:07 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11052347](https://github.com/orgs/community/discussions/22500#discussioncomment-11052347)</b>
    Just to add to the discussion about SMS 2FA.
    Not only is it collecting even more user data, but it is shockingly easy to hack an account with SMS 2FA:
    A simple social engineering attack (for example calling the user's phone company and asking for a new SIM card for that phone number) will easily bypass the security.
    And there are many documented examples of that specific "SIM swap" attack happening in real life. <b>]</b>

\----

<b id="discussioncomment-11052362">[ Explorer09 @ CE 2024-10-25 12:42:40 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11052362](https://github.com/orgs/community/discussions/22500#discussioncomment-11052362)</b>
    "Find who force 2FA to blame."
    GitHub.

    "is the key generation core being not easily copiable from outside."
    Hardware security keys rely on physical security.
    Yes, the key themselves can be cloned if they are physically stolen.
    But remote stealing of security key data is very unlikely (because NFC requires a very close range (a few centimeters) to communicate).
    As long as I physically hold my keys (thus having control on the physical security), there is no security weakness with the keys. <b>]</b>

\----

<b id="discussioncomment-11052638">[ MasterInQuestion @ CE 2024-10-26 10:44:20 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11052638](https://github.com/orgs/community/discussions/22500#discussioncomment-11052638)</b>
    Depends how you use it.
    Eventually fallback to the key problem: key management.

    Note what adversary targets is your access: not necessarily you keys. <b>]</b>

\----

<b id="discussioncomment-11052767">[ Explorer09 @ CE 2024-10-25 13:22:28 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11052767](https://github.com/orgs/community/discussions/22500#discussioncomment-11052767)</b>
    Your argument was about passwords being as strong as the security key.
    While we refute your arguments, you are now dodging questions.

    Don't propose any substitute for security keys while you know little about security keys.
    This whole thread is about "we want security keys as an exclusive second factor; no TOTP", and discussions about password strengths are already off-topic.
    Please don't waste our time. <b>]</b>

\----

<b id="discussioncomment-11053108">[ MasterInQuestion @ CE 2024-10-25 13:49:32 UTC:
[https://github.com/orgs/community/discussions/22500#discussioncomment-11053108](https://github.com/orgs/community/discussions/22500#discussioncomment-11053108)</b>
    I don't use it doesn't mean I don't understand the technicality behind.
    The point was not on keys themselves, but security.

    Note my original "much identical", not "identical".
    I've explained the primary differences.

    On-topic with the adjacent posts.
    Per main topic OT in deed. <b>]</b>
]]

**Comment by b1ackm0x** [comment]
Wow, impressive information exchange. Hard to follow for me at some points.
    I have some beginner questions if you will.

    Regarding 2FA, [FIDO](https://www.google.com/search?hl=en&gl=ca&num=10&q=FIDO+authentication) protocols: what do you think of the Yubico hardware key?
    I believe that someone was able to break it using a sandbox debug (QEMU) recently?

    Actual physical access is usually the weakest link right?
    If so, do you think GPG/SSH key (air-gapped) creation using proper implementation of OpenSSH or GnuPG is secure?
    What other methods you might recommend?
    Hopefully using this method, rolling private keys and storage/physical access would be the weakest link?

    Information would be greatly appreciated, as I mentioned earlier your cryptology knowledge is amazing, kudos.
    Thank you in advance.

**Reply by MasterInQuestion** [reply]
"Yubico hardware key"
    As mentioned, I don't use it.
    "The key benefit of security keys alike: is the key generation core being not easily copiable from outside."
    If it failed this premise (leaving possible "debug" backdoors alike): it gives no extra strength over software implementations. (just hackable the same way)

    However based on your description (sandbox debug with virtual machine):
    I believe in which case the adversary already had access of the key..?
    (haven't researched the exact technique; so might be erroneous)

    The weakest chain, unsure.
    The primary burden of maintaining security is the key management:
    At many times, well regulating the execution environment (workspace).
    The key generation algorithms self could be scarcely exploitable.

**Reply by b1ackm0x** [reply]
What would you recommend as a preferred method, for server management and security algorithms?
    Some that I know and implement, possibly a security risk even socially mentioning which ones if I was a viable target??

    |\*| OpenSSH
    |\*| GnuPG vs. PGP (not open source) the later one being bad IMHO.
    |\*| YubiKey (2FA FIDO2, etc.)
    And preferred VPN or proxy chain (remotely used, GitHub for example)?
    LAN-only when physical access is not available:
    Siphoning Slurping (side-channel exploitations) hopefully is not an issue.

    Oh and salting I suppose, since entropy is the issue and randomness if I'm correct?
    It is not really random?
    As we saw with XZ Utils, MOVEit, CUPS (Common Unix Printing System):
    Where there is a will, there is a way.

    Memorizing 100+ individual keys with decent password strength is daunting.
    Is there anything that you can think of, that would help me add to my repertoire of knowledge?

**Reply by MasterInQuestion** [reply]
For server managements, it much depends on the exact workflow, and expectations.
    In general:
    “Security doesn't itself come by following random security advices or blindly deploying "security-enhanced" setup:
    Well realizing the underlying implementation is the essence to guarantee effectiveness and security.”
    .
    What ensures security is not obscurity but well defined implementation.
    Well known or not doesn't really affect an implementation's security.
    Counting on others' mercy is never a good practice.

    For the proxy choice:
    [ https://github.com/MasterInQuestion/talk/discussions/30#72309e7b-a0f7-4923-b913-c2c5d2218f16](https://github.com/MasterInQuestion/talk/discussions/30#72309e7b-a0f7-4923-b913-c2c5d2218f16)
    ("AB/ABA structure multi-stage")

    If your potential adversary could confiscate your servers outright:
    You need to have some extra considerations.
\
\
    Perfect randomness doesn't exist: it demands the property of infinity. (while which practically unattainable; or doesn't exist)
    All finite sequences must be deterministic. That is, known, certain.

    Salting or not, if the input is already random-like (e.g. "/dev/random"):
    Salting won't make any real difference. (somewhat even, fancy no-op)
    Salting specifically mitigates the case where the input is easily guessable. (which is somewhat unsolvable; per the weakest-chain principle)

    As for the various exploitations, it just met the aforementioned general reference.
    The XZ Utils one further proved why auto-update is bad idea.

    "Where there is a will, there is a way."
    However the price does matter.
    Many things effectively working in such manner:
    Truly unbreakable? No.
    Just expensive enough to deter breakage.
    The art of balancement.

    "Memorizing 100+ individual keys ..."
    Writing the credentials in Plain Text and keep it safe.
    The crucial part is "keep it safe": essentially "well regulating the execution environment (workspace)" aforementioned.

**Reply by b1ackm0x** [reply]
Very well stated, appreciate the info, layout and forum. 🙏

**Comment by b1ackm0x** [comment]
I'm curious, was my post edited for grammar?

Is this an AI exchange, Master & Explorer?

Am I just overthinking this?

In review, I don't think my grammar is, that great?

**Reply by MasterInQuestion** [reply]
For ease of access, generally.
    I indeed use AI techs for linguistic analysis. And content generation, partly.
    (all posted contents are extraordinarily verified for quality)

    You may check the diff from the edit history.
    (probably you have to dump the content and use more effective diff, for certain cases)

    The thread gather is transcribed from corresponding links.
    (some sources might be dead/deleted; the transcription keeps when last accessed)

**Comment by b1ackm0x** [comment]
Awesome we love it full transparency model, thank you..
